plugins/telescope/default.nix

```diff
@@ -59,78 +69,86 @@ in {
-    extraPackages = [pkgs.bat];
+    callSetup = false;
+    extraConfig = cfg: {
+      extraConfigVim = mkIf (cfg.highlightTheme != null) ''
+        let $BAT_THEME = '${cfg.highlightTheme}'
+      '';
+
+      keymaps =
+        mapAttrsToList
+        (
+          key: mapping: let
+            actionStr =
+              if isString mapping
+              then mapping
+              else mapping.action;
+          in {
+            mode = mapping.mode or "n";
+            inherit key;
+            action.__raw = "require('telescope.builtin').${actionStr}";
+
+            options =
+              {
+                silent = cfg.keymapsSilent;
+              }
+              // (mapping.options or {});
+          }
+        )
+        cfg.keymaps;
+
+      extraConfigLua = ''
+        require('telescope').setup(${helpers.toLuaObject cfg.settings})
 
-    extraPlugins = with pkgs.vimPlugins; [
-      cfg.package
-      plenary-nvim
-      popup-nvim
-    ];
-
-    extraConfigVim = mkIf (cfg.highlightTheme != null) ''
-      let $BAT_THEME = '${cfg.highlightTheme}'
-    '';
-
-    keymaps =
-      mapAttrsToList
-      (
-        key: action: let
-          actionStr =
-            if isString action
-            then action
-            else action.action;
-          actionProps =
-            if isString action
-            then {}
-            else filterAttrs (n: v: n != "action") action;
-        in {
-          mode = "n";
-          inherit key;
-          action.__raw = "require('telescope.builtin').${actionStr}";
-
-          options =
-            {
-              silent = cfg.keymapsSilent;
-            }
-            // actionProps;
-        }
-      )
-      cfg.keymaps;
-
-    extraConfigLua = let
-      options =
-        {
-          extensions = cfg.extensionConfig;
-          inherit (cfg) defaults;
-        }
-        // cfg.extraOptions;
-    in ''
-      do
         local __telescopeExtensions = ${helpers.toLuaObject cfg.enabledExtensions}
-
-        require('telescope').setup(${helpers.toLuaObject options})
-
         for i, extension in ipairs(__telescopeExtensions) do
           require('telescope').load_extension(extension)
         end
-      end
-    '';
-  };
-}
+      '';
+    };
+
+    settingsOptions = {
+      defaults = helpers.mkNullOrOption (with types; attrsOf anything) ''
+        Default configuration for telescope.
+      '';
+
+      pickers = helpers.mkNullOrOption (with types; attrsOf anything) ''
+        Default configuration for builtin pickers.
+      '';
+
+      extensions = mkOption {
+        type = with types; attrsOf anything;
+        default = {};
+        # NOTE: We hide this option from the documentation as users should use the top-level
+        # `extensions.*` options.
+        # They can still directly change this `settings.adapters` option.
+        # In this case, they are responsible for explicitly installing the manually added extensions.
+        visible = false;
+      };
```

Comment:
Could this be visible to mitigate issues like #1218?

> Perhaps `plugins.telescope.enabledExtensions` could be made public or an `plugins.telescope.extraExtensions` could be introduced to handle other unsupported telescope extensions?

Or is there a better solution?

Reply:
wouldn't `plugins.telescope.settings.extensions` work for that? It's the main adavantage of RFC 42 style plugins, you should _not_ be limited to what nixvim provides